bpf-sys/src/type_gen.rs

```diff
@@ -0,0 +1,253 @@
+impl Drop for StreamWrapper {
+    fn drop(&mut self) {
+        unsafe {
+            libc::fclose(self.0);
+        }
+    }
+}
+
+// only used for RAII
+struct BtfDumpWrapper(*mut btf_dump);
+impl Drop for BtfDumpWrapper {
+    fn drop(&mut self) {
+        unsafe {
+            btf_dump__free(self.0);
+        }
+    }
+}
+
+/// An error that occurred during parsing vmlinux and generating C source code
+/// from BTF.
+#[derive(Debug)]
+pub enum TypeGenError {
+    /// error on parsing vmlinux
+    VmlinuxParsingError,
+    /// vmlinux not found
+    VmlinuxNotFound,
+    /// path contains invalid utf-8
+    InvalidPath,
+    /// IO error
+    IO(io::Error),
+    /// invalid regex
+    RegexError,
+    DumpError,
+}
+
+type Result<T> = std::result::Result<T, TypeGenError>;
+
+/// Load vmlinux from file and generate `vmlinux.h`
+pub struct VmlinuxBtfDump {
+    whitelist: Option<Vec<String>>,
+    btfptr: *mut btf,
+}
+
+impl VmlinuxBtfDump {
+    /// Probe few well-known locations for vmlinux kernel image and try to load
+    /// BTF data out of it
+    pub fn with_system_default() -> Result<Self> {
+        let btfptr = unsafe { libbpf_find_kernel_btf() };
+        if (btfptr as isize) < 0 {
+            return Err(TypeGenError::VmlinuxNotFound);
+        }
+
+        Ok(VmlinuxBtfDump {
+            whitelist: None,
+            btfptr,
+        })
+    }
+
+    /// Try to load BTF data out of the given ELF file
+    pub fn with_elf_file(elf_file: impl AsRef<Path>) -> Result<Self> {
+        if !elf_file.as_ref().exists() {
+            return Err(TypeGenError::VmlinuxNotFound);
+        }
+
+        let elf_str = elf_file
+            .as_ref()
+            .to_str()
+            .ok_or(TypeGenError::InvalidPath)?;
+
+        let cvmlinux_path = CString::new(elf_str).unwrap();
+        let btfptr = unsafe { btf__parse_elf(cvmlinux_path.as_ptr(), ptr::null_mut()) };
+        if (btfptr as isize) < 0 {
+            return Err(TypeGenError::VmlinuxParsingError);
+        }
+
+        Ok(VmlinuxBtfDump {
+            whitelist: None,
+            btfptr,
+        })
+    }
+
+    /// Add regex `pattern` into whitelist of BTF types
+    pub fn whitelist(mut self, pattern: &str) -> Self {
+        let whitelist: &mut Vec<String> = if let Some(whitelist) = &mut self.whitelist {
+            whitelist
+        } else {
+            self.whitelist = Some(vec![]);
+            self.whitelist.as_mut().unwrap()
+        };
+        whitelist.push(pattern.to_string());
+        self
+    }
+
+    /// Dump BTF types as C source code into `outfile` including all the
+    /// necessary dependent types.
+    pub fn generate(self, outfile: impl AsRef<Path>) -> Result<()> {
+        let path_str = outfile.as_ref().to_str().ok_or(TypeGenError::InvalidPath)?;
+        let cpath = CString::new(path_str).unwrap();
+        let cmode = CString::new("w").unwrap();
+        let stream = unsafe { libc::fopen(cpath.as_ptr(), cmode.as_ptr()) };
```

Comment:
I understand this is a ported code from C, but is there a reason you're using C APIs instead of the `std::fs` equivalent from Rust? I feel these would be low hanging fruits, and not particularly hard to adapt.

Reply:
https://github.com/libbpf/libbpf/blob/506a544834573905ada61da8e00f54b04f0caf43/src/btf.h#L152

libbpf requires `btf_dump_printf_fn_t` so I defined `vfprintf_wrapper` that
directly calls `vfprintf`. Without `vfprintf`, `vfprintf_wrapper` should deal
with `va_list`. But I felt it is tricky.

And `vfprintf` requires `FILE*` so I provided it by calling `libc::fopen`.

I wanted to use `std::fs` but I din't know whether it provides a function returning `FILE *`.

Reply:
[`File::into_raw_fd()`](https://doc.rust-lang.org/std/fs/struct.File.html#method.into_raw_fd) helps you get the underlying fd, and you can use `vdprintf` with that fd to achieve the same results. There's also a [crate](https://crates.io/crates/va_list) that could help with dealing with variadic arguments, so you can keep most of this purely rust (e.g. get the raw pointer (_not_ fd) to the `File` instance if you know it outlives the callback, and convert it back on the other side with `unsafe`, then do something rusty to write the data to the file).

Reply:
It is a good idea. I will try to fix it. Thanks,